typing/includemod_errorprinter.ml

```diff
@@ -95,43 +95,50 @@ module Illegal_permutation = struct
     | None -> g y
     | Some _ as v -> v
 
+  type change =
+    | Transposition of int * int
+    | Primitive_coercion of string
+    | Alias_coercion of Path.t
+
   (** We extract a lone transposition from a full tree of permutations. *)
   let rec transposition_under path (coerc:Typedtree.module_coercion) =
     match coerc with
     | Tcoerce_structure(c,_) ->
-        either
-          (not_fixpoint path 0) c
-          (first_non_id path 0) c
+          (either
+             (not_fixpoint path 0) c
+             (first_non_id path 0) c
+          )
     | Tcoerce_functor(arg,res) ->
         either
           (transposition_under (InArg::path)) arg
           (transposition_under (InBody::path)) res
     | Tcoerce_none -> None
-    | Tcoerce_alias _ | Tcoerce_primitive _ ->
-        (* these coercions are not inversible, and raise an error earlier when
-           checking for module type equivalence *)
-        assert false
+    | Tcoerce_alias _ | Tcoerce_primitive _ -> None
+    (* those two coercions cannot happen at toplevel *)
+
   (* we search the first point which is not invariant at the current level *)
   and not_fixpoint path pos = function
     | [] -> None
     | (n, _) :: q ->
-        if n = pos then
+        if n < 0 || n = pos then
```

Comment:
Could we have a comment explaining the `n < 0` case ?
My understanding is that before this patch, we couldn't reach the `n < 0` case because it would have caused another error earlier.
With this patch we also use this code in contexts where the condition can occur, but `n < 0` implies that there is a `Tcoerce_alias` or `Tcoerce_primitive` coercion and we will get a better message error using the `first_non_id` function below.

Reply:
Indeed, the previous state of the code was using the invariant that it was only called on equivalent module types. Thus the coercion was necessarily a permutation of components.

Maybe

> (* if n < 0, this is not a permutation but a kind coercion, which will be covered in the `first_non_id` case *)

?

Reply:
I don't know how it reads for someone familiar with this part of the code, but I spent some time wondering why we did not report this as an error before eventually managing to find that `not_fixpoint` runs before `first_non_id` and so we would get worse messages if we did not skip this case.
So I would prefer if you could manage to fit somewhere in the comment that `n < 0` is not a fixpoint (or rename `not_fixpoint` to `find_permutation`, this way your proposed comment becomes very clear).

Reply:
Hm, there are too many negative here: `n<0` is not a transposition (aka not not-a-fixpoint in the permutation). But this is indeed a sign that the names should be improved: `find_transposition` should be clearer (probably even for people that don't have the definition of transposition in mind?).

Reply:
That works for me, indeed. Although the functions `transposition` and `transposition_under` now return a `change` that could be a transposition or not, so maybe a few more names need to be adapted.
If you don't want to spend too much time on this, you can also just adopt the comment in your earlier suggestion and leave further cleanup to an hypothetical future PR.

Reply:
I have changed few more names to better reflect that we are now extracting a small atomic change from any coercion rather than just a transposition from a permutation.